File: backend/app/api/labels.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.engine import get_db
from app.schemas.label import LabelCreate, LabelListResponse, LabelOut
from app.services.label import LabelService

router = APIRouter()


@router.get("/labels", response_model=LabelListResponse)
def list_labels(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: str | None = Query(None),
    sort_by: str = Query("name"),
    sort_dir: Literal["asc", "desc"] = Query("asc"),
    db: Session = Depends(get_db),
):
    service = LabelService(db)
    return service.list_labels(
        page=page, page_size=page_size, q=q, sort_by=sort_by, sort_dir=sort_dir
    )


@router.post("/labels", response_model=LabelOut, status_code=201)
def create_label(data: LabelCreate, db: Session = Depends(get_db)):
    service = LabelService(db)
    return service.create_label(data)


@router.delete("/labels/{label_id}", status_code=204)
def delete_label(label_id: int, db: Session = Depends(get_db)):
    service = LabelService(db)
    deleted = service.delete_label(label_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Label not found")

File: frontend/src/api/labels.ts
import { api } from "./client";

export interface Label {
  id: number;
  name: string;
  color: string;
  created_at: string;
}

export interface LabelListResponse {
  items: Label[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListLabelsParams {
  page?: number;
  page_size?: number;
  q?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
}

export function listLabels(params?: ListLabelsParams): Promise<LabelListResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.page_size) search.set("page_size", String(params.page_size));
  if (params?.q) search.set("q", params.q);
  if (params?.sort_by) search.set("sort_by", params.sort_by);
  if (params?.sort_dir) search.set("sort_dir", params.sort_dir);
  const qs = search.toString();
  return api.get<LabelListResponse>(`/api/labels${qs ? `?${qs}` : ""}`);
}

export function createLabel(data: { name: string; color?: string }): Promise<Label> {
  return api.post<Label>("/api/labels", data);
}

export function deleteLabel(id: number): Promise<void> {
  return api.delete<void>(`/api/labels/${id}`);
}
